### DISCUSSION TITLE
riscv32-unknown-elf-gcc mixes riscv system includes with host I86-64 system includes

### DISCUSSION THREAD

**Discussion by sylvainhuard** [stnolting/neorv32#686] [Q&A] [answered]
I'm having a system library issue.
I try to compile a software example from the NEORV32 sw/example. The terminal command is simple. It is using the software framework coming from NEORV32 with no motification, straight from GIT.

`~/neorv32-setups/neorv32/sw/example/demo_slink$ make image`

The result is a fatal error but what is difficult to understand is the first few system includes are from the right location escept "sys/cdefs.h" using the path from my host GCC (Ubuntu 22.04 on Intel64). I added to the make file a line before the compiler call to make sure of which GCC was called. It effectively the riscv gcc. See the result:
```
riscv32-unknown-elf-gcc
In file included from /home/shuard/fpgatool/opt/riscv/riscv32-unknown-elf/include/limits.h:5,
                 from /home/shuard/fpgatool/opt/riscv/lib/gcc/riscv32-unknown-elf/12.1.0/include-fixed/limits.h:203,
                 from /home/shuard/fpgatool/opt/riscv/lib/gcc/riscv32-unknown-elf/12.1.0/include-fixed/syslimits.h:7,
                 from /home/shuard/fpgatool/opt/riscv/lib/gcc/riscv32-unknown-elf/12.1.0/include-fixed/limits.h:34,
                 from /home/shuard/neorv32-setups/neorv32/sw/lib/include/neorv32.h:53,
                 from main.c:41:
/usr/include/x86_64-linux-gnu/sys/cdefs.h:24:11: fatal error: features.h: No such file or directory
   24 | # include <features.h>
      |           ^~~~~~~~~~~~
compilation terminated.
make: *** [/home/shuard/neorv32-setups/neorv32/sw/common/common.mk:186: main.c.o] Error 1

```
Running gcc configuration verification show that everything should be OK in terms of include header paths. I'm specifically looking for the 2 varaibles : --with-native-system-header-dir=/include --with-native-system-header-dir=/include  showing the right paths. 
 
```
~/neorv32-setups/neorv32/sw/example/demo_slink$ riscv32-unknown-elf-gcc -v
Using built-in specs.
COLLECT_GCC=riscv32-unknown-elf-gcc
COLLECT_LTO_WRAPPER=/home/shuard/fpgatool/opt/riscv/bin/../libexec/gcc/riscv32-unknown-elf/12.1.0/lto-wrapper
Target: riscv32-unknown-elf
Configured with: /tmp/rv_gcc/riscv-gnu-toolchain/gcc/configure --target=riscv32-unknown-elf --prefix=/opt/riscv --disable-shared --disable-threads --enable-languages=c,c++ --with-pkgversion=g1ea978e3066 --with-system-zlib --enable-tls --with-newlib --with-sysroot=/opt/riscv/riscv32-unknown-elf --with-native-system-header-dir=/include --disable-libmudflap --disable-libssp --disable-libquadmath --disable-libgomp --disable-nls --disable-tm-clone-registry --src=/tmp/rv_gcc/riscv-gnu-toolchain/gcc --disable-multilib --with-abi=ilp32 --with-arch=rv32i --with-tune=rocket --with-isa-spec=2.2 'CFLAGS_FOR_TARGET=-Os   -mcmodel=medlow' 'CXXFLAGS_FOR_TARGET=-Os   -mcmodel=medlow'
Thread model: single
Supported LTO compression algorithms: zlib
gcc version 12.1.0 (g1ea978e3066) 
```
So how is GCC use the right include paths for several includes and suddenly decide to use the host system include path?
I installed the Precompiled toolchain suggested:
 
[https://github.com/stnolting/riscv-gcc-prebuilt](url)

Any hint will be appreciated. I might have just made a stupid mistake or may be there is a bug. I spent several hours looking at everything I would thing of. I've used successfully multiple GCC cross compiler over the years and this the first time I see mixing include folders from multiple versions in the same compile run. 

Thanks,
S.Huard

**Comment by sylvainhuard** [comment, marked as answer]
I found the Issue. I publish it in case someone else face the same problem. On my workstation, there is a CPATH environment variable CPATH. This is the culprit. If a remove it, it works. It's a bit annoying since it might be required for another of my tools. I guess I will figure it out next time I use the tool that needs it.

**Comment by stnolting** [comment]
I have never encountered this... However, thank you very much for sharing your finding! :+1:
